internal/pkg/agent/application/upgrade/rollback_test.go

```diff
@@ -0,0 +1,507 @@
+		versionedHome = fmt.Sprintf("elastic-agent-%s-%s", version, hash[:hashLen])
+	}
+	relVersionedHomePath := filepath.Join("data", versionedHome)
+	absVersionedHomePath := filepath.Join(topDir, relVersionedHomePath)
+
+	// recalculate the binary path and launch a mkDirAll to account for MacOS weirdness
+	// (the extra nesting of elastic agent binary within versionedHome)
+	absVersioneHomeBinaryPath := paths.BinaryPath(absVersionedHomePath, "")
+	err := os.MkdirAll(absVersioneHomeBinaryPath, 0o750)
+	require.NoError(t, err, "error creating fake install versioned home directory (including binary path) %q", absVersioneHomeBinaryPath)
+
+	// place a few directories in the fake install
+	absComponentsDirPath := filepath.Join(absVersionedHomePath, "components")
+	err = os.MkdirAll(absComponentsDirPath, 0o750)
+	require.NoError(t, err, "error creating fake install components directory %q", absVersionedHomePath)
+
+	absLogsDirPath := filepath.Join(absVersionedHomePath, "logs")
+	err = os.MkdirAll(absLogsDirPath, 0o750)
+	require.NoError(t, err, "error creating fake install logs directory %q", absLogsDirPath)
+
+	absRunDirPath := filepath.Join(absVersionedHomePath, "run")
+	err = os.MkdirAll(absRunDirPath, 0o750)
+	require.NoError(t, err, "error creating fake install run directory %q", absRunDirPath)
+
+	// put some placeholder for files
+	agentExecutableName := agentName
+	if runtime.GOOS == "windows" {
+		agentExecutableName += ".exe"
+	}
+	err = os.WriteFile(paths.BinaryPath(absVersionedHomePath, agentExecutableName), []byte(fmt.Sprintf("Placeholder for agent %s", version)), 0o750)
+	require.NoErrorf(t, err, "error writing elastic agent binary placeholder %q", agentExecutableName)
+	err = os.WriteFile(filepath.Join(absLogsDirPath, "fakelog.ndjson"), []byte(fmt.Sprintf("Sample logs for agent %s", version)), 0o750)
+	require.NoErrorf(t, err, "error writing fake log placeholder %q")
+
+	// return the path relative to top exactly like the step_unpack does
+	return relVersionedHomePath
+}
+
+func createLink(t *testing.T, topDir string, currentAgentVersionedHome string) {
+	linkTarget := paths.BinaryPath(currentAgentVersionedHome, agentName)
+	linkName := agentName
+	if runtime.GOOS == "windows" {
+		linkTarget += ".exe"
+		linkName += ".exe"
+	}
+	err := os.Symlink(linkTarget, filepath.Join(topDir, linkName))
+	require.NoError(t, err, "error creating test symlink to fake agent install")
+}
+
+func createUpdateMarker(t *testing.T, log *logger.Logger, topDir, newAgentVersion, newAgentHash, newAgentVersionedHome, oldAgentVersion, oldAgentHash, oldAgentVersionedHome string, useNewMarker bool) {
+
+	if !useNewMarker {
+		newAgentVersion = ""
+		newAgentVersionedHome = ""
+		oldAgentVersionedHome = ""
+	}
+
+	err := markUpgrade(log, paths.DataFrom(topDir), newAgentVersion, newAgentHash, newAgentVersionedHome, oldAgentVersion, oldAgentHash, oldAgentVersionedHome, nil, nil)
+	require.NoError(t, err, "error writing fake update marker")
```

Comment:
This always creates the marker with the new fields you added, in the case where we come from an old agent, shouldn't it omit them to make the test 100% accurate?

Reply:
there is a boolean parameter for that in `createUpdateMarker` that blanks out the new fields to simulate an old upgrade marker just a few lines above this comment
https://github.com/elastic/elastic-agent/pull/4176/files/bf2e849b16f8852a59593e45b8ada9095e190ad1..b17c7132e2b6cd2ff157835379c833435f325fc4#diff-cc4058ffab0db853a72dc782e00a1a7c3c7b7314175bec4538b1f32e69332c66R497-R503